### ISSUE TITLE
syncserver error handling should be better

### ISSUE THREAD

**Issue by hagedose** [cyrusimap/cyrus-imapd#2120]
I have a production server running 2.4.20 and a testing server with 3.0.3. The production server has many partitions, whereas the testing server currently only has the default partition.

I created my account on the testing server and appended a few messages to the INBOX. Later I wanted to experiment with replication and first tried to manually sync my inbox like this:

```
sync_client -vm user/a0620
MAILBOXES user.a0620
Error from do_mailboxes(): bailing out!
```

On the testing server I saw this in the log:

```
Aug 25 00:39:56 cyrus3-test cyrus/syncserver[17457]: conversations_rename_folder: deleted user.a0620
Aug 25 00:39:56 cyrus3-test cyrus/syncserver[17457]: Deleted mailbox user.a0620
Aug 25 00:40:19 cyrus3-test cyrus/syncserver[17457]: Fatal error: Internal error: assertion failed: imap/dlist.c: 156: base != NULL

```
First, I found it a little surprising that the existing mailbox is simply deleted. Then the error message isn't very helpful unless you’re familiar with server internals. What's more, afterwards there was a broken entry in mailboxes.db:

```
# ctl_mboxlist -d
user.a0620	16 (null)
```

I figured out that the underlying problem is that my mailbox is on a partition that doesn't exist on the testing server. Could the error handling for this scenario be improved? Is the requirement for partitions to be identical on all replica servers even documented? I might have overlooked it, but I don't recall seeing it.

**Comment by hagedose** [comment]
FWIW, I just reproduced the issue with 3.0.5:

```
Jan 04 15:32:41 cyrus3-test.rrz.uni-koeln.de cyrus/syncserver[3265]: conversations_rename_folder: deleted user.a0620
Jan 04 15:32:41 cyrus3-test.rrz.uni-koeln.de cyrus/syncserver[3265]: Deleted mailbox user.a0620
Jan 04 15:32:41 cyrus3-test.rrz.uni-koeln.de cyrus/imap[3169]: graceful shutdown initiated by unexpected process 3265 (sync_server)
Jan 04 15:32:41 cyrus3-test.rrz.uni-koeln.de cyrus/master[3146]: process type:SERVICE name:imap path:/usr/lib/cyrus-imapd/imapd age:768.769s pid:3169 exited, status 75
Jan 04 15:32:41 cyrus3-test.rrz.uni-koeln.de cyrus/master[3146]: service imap/ipv6 pid 3169 in BUSY state: terminated abnormally
Jan 04 15:32:41 cyrus3-test.rrz.uni-koeln.de cyrus/syncserver[3265]: Fatal error: Internal error: assertion failed: imap/dlist.c: 156: base != NULL
Jan 04 15:32:41 cyrus3-test.rrz.uni-koeln.de cyrus/master[3146]: process type:SERVICE name:syncserver path:/usr/lib/cyrus-imapd/sync_server age:0.607s pid:3265 exited, status 75
Jan 04 15:32:41 cyrus3-test.rrz.uni-koeln.de cyrus/master[3146]: service syncserver/ipv4 pid 3265 in BUSY state: terminated abnormally
```

**Comment by elliefm** [comment]
I thought that requirement was documented, but I don't recall where.  Maybe I figured it out from code and then misremembered it as being documented?

**Comment by elliefm** [comment]
Here's a patch against master, but which should also apply cleanly to 3.0.5.  It adds syslog warnings when trying to look up a partition which hasn't been defined.

I expect that it would still Fatal out at the same spot, but at least a more useful reason why should be logged too.  I'm not sure how feasible it is to change the behaviour more than this -- we're down a pretty deep rabbit hole by the time the problem becomes detectable.

Still to do here: find a suitable spot in the docs for the "partitions must match" detail, and document it.

**Comment by hagedose** [comment]
Thanks! I can confirm that with the patch applied there are syslog warnings:

```
Jan 05 11:45:03 cyrus3-test.rrz.uni-koeln.de cyrus/syncserver[28582]: requested archive partition directory for unknown partition 'default'
Jan 05 11:45:03 cyrus3-test.rrz.uni-koeln.de cyrus/syncserver[28582]: requested partition directory for unknown partition 'imap2'
Jan 05 11:45:03 cyrus3-test.rrz.uni-koeln.de cyrus/syncserver[28582]: Fatal error: Internal error: assertion failed: imap/dlist.c: 156: base != NULL

```
The partition 'default' does exist, but there is no archive partition for it. The partition 'imap2' does not exist at all.

I wonder if the logging for partitions other than the data partition isn't overkill. Archive and meta partitions are optional ...

I'm still not happy that existing data is deleted and that after the failure there is this entry in mailboxes.db, but I guess if the requirement for exactly matching partitions on source and target server is documented I can live with that:

`user.a0620	16 (null)`

**Comment by hagedose** [comment]
Regarding where to document the requirement, it should be here:

https://www.cyrusimap.org/imap/reference/admin/sop/replication.html

I'd suggest putting a "**Note**" or even an "**Important**" in the "Replica server configuration" section.

**Comment by elliefm** [comment]
> I wonder if the logging for partitions other than the data partition isn't overkill. Archive and meta partitions are optional ...

Yeah I agree with this -- we shouldn't be logging them as LOG_WARNING when it's not a problem for them to be missing.  I'll downgrade these two to LOG_DEBUG, merge this to master and 3.0, and also backport @postilion's documentation update to 3.0.

**Comment by elliefm** [comment]
Work so far is now on master and 3.0 branches.

I'm wondering if there's more we can do to catch and report this before we choke on it, though.  It's Fataling out at the stage where the messages are being uploaded to the mailbox, but maybe it should have detected the impossibility of the situation and returned an error earlier.

```
sync_client -vm user/a0620
MAILBOXES user.a0620
Error from do_mailboxes(): bailing out!
```

Could you possibly run this again but with three "v"'s, e.g. `sync_client -vvvm user/a0620`  (this is the highest level of verbosity that means anything to sync_client, and will show the raw sync protocol talk) and paste the output?  Or email the output directly to me, ellie@fastmail.com, if it's unsuitable for posting to github.  I'd like to see exactly what it's doing.

**Comment by hagedose** [comment]
That didn't do anything:

```
sync_client -vvv -m user/a0620
MAILBOXES user.a0620
Error from do_mailboxes(): bailing out!

```
Maybe this additional syslog logging helps?

```
Jan  9 11:41:57 xxx sync_client[8949]: MESSAGE received * response:
Jan  9 11:41:57 xxx sync_client[8949]: do_folders(): update failed: user.a0620 'Bad protocol'
Jan  9 11:41:57 xxx sync_client[8949]: Error in do_mailboxes(): bailing out!

```

**Comment by elliefm** [comment]
Oh wait, you're syncing from 2.4... I guess a sync_client that old doesn't support the extra verbosity levels, doh!  That syslog doesn't really help either unfortunately.  Thanks for trying though. :)

I can probably get by without it, just means I have to do more work on this end to figure out what our options are! :)  The plus side of doing the extra work is that we'll probably get a free regression test out of it for later, so I'll just get on with it.

**Comment by elliefm** [comment]
> `user.a0620 16 (null)`

I think we worked out on the mailing list a while ago that this thing is a tombstone record (the 16 == MBTYPE_DELETED), which is a new feature on 3.0.  #2265 is currently open to get it documented.

So I think the main piece of work remaining here is detecting the case where a mailbox that exists on a replication master is on a different partition on the replica, and complaining _without_ deleting the replica copy.  Does that sound right to you?

**Comment by hagedose** [comment]
Yes, I think that would be the ideal scenario.
